### PR TITLE
Optimise transaction preverification

### PR DESCRIPTION
**Pull request by td202**
## Purpose

Addresses #596.

This changes the handling of transactions that are received apart from a block so that the verification step occurs before the global state write lock is acquired. This reduces the contention for the global state lock.

## Changes

- `addVerifiedTransaction` is added to the `TreeState` monad. It's behaviour is similar to `addCommitTransaction`, but it does not perform the verification of the transaction, instead receiving the (successful) result as an argument.
- `preverifyTransaction` is added to `SkovQueryMonad`. This performs initial verification of a transaction without updating the state.
- `addPreverifiedTransaction` is added to `SkovMonad`. This adds a preverified transaction to the state with `addVerifiedTransaction`. Together, `preverifyTransaction` and `addPreverifiedTransaction` perform the same task as `receiveTransaction` (from `SkovMonad`).
- `Concordium.MultiVersion.receiveTransaction` is changed to first call `preverifyTransaction` without acquiring the global state write lock, and then call `addPreverifiedTransaction` while holding the global state lock, if the verification is successful.
- `transactionVerificationResultToUpdateResult` is moved from `Concordium.Skov.Update` to `Concordium.Skov.Monad` as it is now used in `Concordium.MultiVersion.receiveTransaction`.

## Checklist

- [X] My code follows the style of this project.
- [ ] The code compiles without warnings.
- [X] I have performed a self-review of the changes.
- [X] I have documented my code, in particular the intent of the
      hard-to-understand areas.
- [ ] (If necessary) I have updated the CHANGELOG.

### CODE REVIEW

**Review by EmilLa1 — APPROVED** [review]
Looks great! Minor comments that are of no big importance.

File: concordium-consensus/src/Concordium/Skov/MonadImplementations.hs [hunk 1]
```diff
@@ -600,7 +601,8 @@ instance
     storeBlock = doStoreBlock
 
     {- - INLINE receiveTransaction - -}
-    receiveTransaction tr = doReceiveTransaction tr
+    receiveTransaction = doReceiveTransaction
+    addPreverifiedTransaction = doAddPreverifiedTransaction
```
[the remark this answers is not in the document]
Comment (td202) [reply concordium-consensus/src/Concordium/Skov/MonadImplementations.hs:597]: The other INLINEs are commented out, so I don't think it makes sense to add an inline for this. Instead, I've removed commented ones, as they are not adding value.

File: concordium-consensus/src/Concordium/Skov/Update.hs [hunk 2]
```diff
@@ -672,6 +672,43 @@ doReceiveTransaction tr = unlessShutDown $ do
     when (ur == ResultSuccess) $ purgeTransactionTable False =<< currentTime
     return ur
 
+-- |Add a transaction that has previously been verified, given the result of verification.
+doAddPreverifiedTransaction ::
+    (TreeStateMonad m) =>
+    BlockItem ->
+    TV.OkResult ->
+    m UpdateResult
+doAddPreverifiedTransaction bi0 okRes = do
```
[the remark this answers is not in the document]
Comment (td202) [reply concordium-consensus/src/Concordium/Skov/Update.hs]: OK.

File: concordium-consensus/src/Concordium/GlobalState/TransactionTable.hs [hunk 3]
```diff
@@ -226,6 +226,35 @@ emptyTransactionTableWithSequenceNumbers accs upds =
           _ttNonFinalizedChainUpdates = emptyNFCUWithSequenceNumber <$> Map.filter (/= minUpdateSequenceNumber) upds
         }
 
+-- |Add a transaction to a transaction table if its nonce/sequence number is at least the next
+-- non-finalized nonce/sequence number.  A return value of 'True' indicates that the transaction
+-- was added.  The caller should check that the transaction is not already present.
+addTransaction :: BlockItem -> Slot -> TVer.VerificationResult -> TransactionTable -> (Bool, TransactionTable)
+addTransaction blockItem@WithMetadata{..} slot verRes tt0 =
```
Comment (abizjak) [inline concordium-consensus/src/Concordium/GlobalState/TransactionTable.hs:233]: This should probably be strict in the `verRes`.

**Review by abizjak — APPROVED** [review]

**Comment by abizjak** [comment]
The changelog should be updated to reflect this change.

**Pull request merged**